### PR TITLE
Try using `Fin` instead of inductive type for referring to variables

### PR DESCRIPTION
**Pull request by tobiasgrosser**
This should decrease the term size `simp` has to traverse, but it makes elaboration more expensive and does not seem to work if the context is not fully known

### CODE REVIEW

File: SSA/Projects/InstCombine/InstCombineSlow.lean [hunk 1]
```diff
@@ -334,28 +335,19 @@ theorem alive_constant_fold_10: forall (w : Nat), TSSA.eval
...
-  %v151 := op:add w %v150;
-  %v160 := pair:%v151 %v001;
-  %v161 := op:add w %v160;
-  %v170 := pair:%v161 %v001;
-  %v171 := op:add w %v170;
-  %v180 := pair:%v171 %v001;
-  %v181 := op:add w %v180;
-  %v190 := pair:%v181 %v001;
-  %v191 := op:add w %v190
-  dsl_ret %v191
```
Comment (bollu) [inline SSA/Projects/InstCombine/InstCombineSlow.lean:348]: Why were these removed?

File: SSA/Core/WellTypedFramework.lean [hunk 2]
```diff
@@ -207,19 +212,18 @@ def TSSA.eval {Op β : Type} [Goedel β] [TypedUserSemantics Op β] :
   {Γ : Context β} → {i : TSSAIndex β} → TSSA Op Γ i → (e : EnvC Γ) → i.eval
 | Γ, _, .assign lhs rhs rest, e =>
     rest.eval (fun
-      | _, Context.Var.prev v => e v
-      | _, Context.Var.last => rhs.eval e)
+      | _, { idx := ⟨i + 1, h⟩, of_type } => e { idx := ⟨i, Nat.lt_of_succ_lt_succ h⟩, of_type }
```
Comment (bollu) [inline SSA/Core/WellTypedFramework.lean:215]: I wonder if we can make this look cleaner using `@[match_pattern]` to create a pattern synonym for what we are matching here. The current match is quite a whirlwind of syntax.

**Review by bollu — COMMENTED** [review]
Do we have benchmark numbers that tell us how the performance changes?

I would prefer we merge this after we (a) have numbers and (b) try to make the match look nicer. However, since I'm not actively available, I shall not block the merge.

**Comment by tobiasgrosser** [comment]
We are currently still playing with the encoding. This seems -- so far -- to be a failed attempt.

**Comment by tobiasgrosser** [comment]
This is outdated. Close.
